ICML-Submission/saliency.py

```diff
@@ -0,0 +1,170 @@
+import tensorflow as tf
+import numpy as np
+from matplotlib.patches import Ellipse
+import matplotlib.pyplot as plt
+import tensorflow.keras.backend as K
+from tensorflow.keras.layers import Input
+from tensorflow.keras.models import Model
+from tensorflow.keras.activations import softmax
+
+
+# change to vis saliency map
+def VisualizeImageGrayscale(image_3d):
+    '''
+    Returns a 3D tensor as a grayscale 2D tensor by summing the absolute values over the last axis and scaling the results to the interval [0,1].
+    :param image3d: A 3d numpy array of shape (H, W, C).
+    :return: numpy array of shape (H, W) that contains the 2d grey_scale image.
+    '''
+
+    image_2d = np.sum(np.abs(image_3d), axis=2)
+    vmax = np.max(image_2d)
+    vmin = np.min(image_2d)
+    if vmax == vmin:
+        return np.zeros(image_2d.shape)
+    return (image_2d - vmin) / (vmax - vmin)
+
+def get_saliency_maps_and_fitted_ellipses(model, layers_index_list, no_of_nodes, image):
+    gradients = get_gradients_at_layers(model, layers_index_list, no_of_nodes, image)
+    reshaped_grads = np.reshape(gradients, (gradients.shape[0] * gradients.shape[1],) + gradients.shape[2:])

```

Comment:
Add a comment here explaining what this does. Then we don't need a test. So this turns (20, 30, 5, 2, 9) into (600, 5, 2, 9).
Can you comment what shape the gradients have and why this operation is necessary?